cuda/solver/gmres_kernels.cu

```diff
@@ -89,12 +436,126 @@ void step_1(std::shared_ptr<const CudaExecutor> exec,
+        __syncthreads();
+    }
+    // Solve upper triangular.
+    // y = hessenberg \ residual_norms
+}
+
+
+template <size_type block_size, typename ValueType>
+__global__ __launch_bounds__(block_size) void calculate_Qy(
+    const size_type num_rows, const size_type num_cols, const size_type num_rhs,
+    const ValueType *__restrict__ krylov_bases, const ValueType *__restrict__ y,
+    ValueType *__restrict__ before_preconditioner,
+    const size_type *__restrict__ final_iter_nums)
+{
+    constexpr auto warps_per_block = block_size / cuda_config::warp_size;
+    const auto global_id =
+        thread::get_thread_id<cuda_config::warp_size, warps_per_block>();
+    const auto row_id = global_id / num_rhs;
+    const auto col_id = global_id % num_rhs;
+
+    before_preconditioner[global_id] = zero<ValueType>();
+    for (size_type j = 0; j < final_iter_nums[col_id]; ++j) {
+        before_preconditioner[global_id] +=
+            krylov_bases[row_id * num_cols + j * num_rhs + col_id] *
+            y[j * num_rhs + col_id];
+    }
+}
+
+
+}  // namespace kernel
+
+
+template <typename ValueType>
+void solve_upper_triangular(const matrix::Dense<ValueType> *residual_norms,
+                            matrix::Dense<ValueType> *hessenberg,
+                            matrix::Dense<ValueType> *y,
+                            const Array<size_type> *final_iter_nums)
+{
+    // TODO: tune this parameter
+    // TODO: when number of right hand side is larger than block_size
+    constexpr auto block_size = default_block_size;
+    const dim3 block_dim{cuda_config::warp_size, 1,
+                         block_size / cuda_config::warp_size};
+
+    kernel::solve_upper_triangular<block_size><<<1, block_dim>>>(
+        hessenberg->get_size()[1], residual_norms->get_size()[1],
+        as_cuda_type(residual_norms->get_const_values()),
+        as_cuda_type(hessenberg->get_values()), as_cuda_type(y->get_values()),
+        as_cuda_type(final_iter_nums->get_const_data()));
+}
+
+
+template <typename ValueType>
+void solve_x(std::shared_ptr<const CudaExecutor> exec,
+             matrix::Dense<ValueType> *krylov_bases,
+             matrix::Dense<ValueType> *y, matrix::Dense<ValueType> *x,
+             const Array<size_type> *final_iter_nums,
+             const LinOp *preconditioner)
+{
+    auto before_preconditioner =
+        matrix::Dense<ValueType>::create_with_config_of(x);
+    auto after_preconditioner =
+        matrix::Dense<ValueType>::create_with_config_of(x);
+
+    constexpr auto block_size = default_block_size;
+    const dim3 grid_dim =
+        ceildiv(x->get_size()[0] * x->get_size()[1], block_size);
+    const dim3 block_dim{cuda_config::warp_size, 1,
+                         block_size / cuda_config::warp_size};
+
+    kernel::calculate_Qy<block_size><<<grid_dim, block_dim>>>(
+        before_preconditioner->get_size()[0], krylov_bases->get_size()[1],
+        before_preconditioner->get_size()[1],
```

Comment:
`x` and therefore `before_preconditioner` and `after_preconditioner` can have a stride that is different from the number of columns. Thus, you also need the stride as a parameter.